Feature/job generation

micah-wehrle opened the pull request:
Front end PR: https://github.com/StreamSG/ABCWorkshop/pull/36

Admittedly this should have been in master a while ago, I've been using it through quite a few front end PRs. Technically the current master branch here isn't incompatible with the front end, but all the job functionality is contained within this PR.

It's a big one, I'm sorry. But it's pretty simple.
Essentially when the front end requests a job list, it passes a uuid, and optionally the number of jobs to generate. This info is then passed into a model file (which honestly idk if that's technically what a "model" file should be--let's talk about it). The class in the model file will then generate ALL the job data using the uuid as a seed, and then return a fat object which is shot back to the front end.

If you look at the ticket generation, it's relatively straightforward. It just creates an object with most values being filled with a `generate[SomeAspect]()` method. There's one for pretty much everything. 
The order "matters" meaning that as long as the order isn't disturbed, it'll always generate the same jobs for any given uuid.

I'd like to eventually make it day-specific, but don't think that's needed in the short term. Maybe something that someone else would like to try to tackle? I can help with what considerations would go into that (a bit of front end work, as well!)

Let's chop this PR up together. Thanks

Files changed:
`jobs.controller.ts`
Files created (and thoroughly filled):
`tech-load.model.ts`

micah-wehrle committed:
began work on job generation

micah-wehrle committed:
continued job generation

micah-wehrle committed:
changed JobData structure, rename City to Location

micah-wehrle committed:
Made randDigits guarentee length

micah-wehrle committed:
randDigits bugfix

micah-wehrle committed:
Added ability to request specific number of jobs

micah-wehrle committed:
Fixed missing city db data

micah-wehrle committed:
Combined :uuid/ and :uuid/:amount routes

micah-wehrle committed:
Merge branch 'master' of https://github.com/micah-wehrle/ABCNest into feature/jobGeneration

micah-wehrle committed:
finished jsdocs

micah-wehrle commented on src/modules/jobs/models/tech-load.model.ts:
```diff
@@ -0,0 +1,399 @@
+export class TechLoad {
+
+  public jobs: JobData[];
```
do these need to be public? Or even global?

looksliketyler replied:
I don't see `jobs` being used anywhere other that here...?

micah-wehrle replied:
Hmm. It's not used here, but "technically" is used on the front end. Not sure how it packages the data to be sent, but I assume it takes every public variable of the class and then converts it into a generic object. prob json.stringify. I'll have to test, but I'm guessing if I make all of this data private or not global than it will not be sent to the front end.

micah-wehrle replied:
I made them private except for jobs, and changed how I pass the job data to the front end. I was sending this entire class object to the front end, which includes all public and private variables. I instead made everything private except jobs, and then just send the array of jobs to the front end. I adjusted the front end service file accordingly in the sister PR

looksliketyler commented on src/modules/jobs/controllers/jobs.controller.ts:
```diff
@@ -1,14 +1,26 @@
+  async getJobsSpecific(@Param('uuid') uuid: string, @Param('amount') amount: string) {
+    await new Promise(resolve => setTimeout(resolve, 1500)); // just adds a delay for feel
+
+    if (!uuid || uuid.length === 0) {
```
A length of zero would be a falsy value

micah-wehrle replied:
I don't follow

StreamSG replied:
first condition covers both in this case

micah-wehrle replied:
Oh I was thinking this is an array. I gotcha, ok thanks.

looksliketyler commented on src/modules/jobs/models/tech-load.model.ts:
```diff
@@ -0,0 +1,399 @@
+   */
+  private generateJob(): JobData {
+    // The following variables are generated prior to the output object, as they are passed into generator methods in order to generate data consistent throughout the ticket
+    const firstName = this.generateFirstName();
```
Missing data types for these consts

looksliketyler commented on src/modules/jobs/models/tech-load.model.ts:
```diff
@@ -0,0 +1,399 @@
+    const jobType = this.randOf(['Install', 'Repair', 'Helper', 'POTS', 'BSW']);
+    const transportType = jobType === 'POTS' || this.nextRand() < 0.5 ? this.randOf(['IP-CO', 'IP-RT', 'FTTN', 'FTTN-BP']) : this.randOf(['FTTP GPON', 'FTTP XGS-PON']);
+    const curJob: JobData = {
+      firstName: firstName,
```
You could just put `firstName,`. Such with last name

micah-wehrle replied:
This seems like something Atlas wouldn't like but I changed it.

looksliketyler commented on src/modules/jobs/models/tech-load.model.ts:
```diff
@@ -0,0 +1,399 @@
+    let hash = 0;
+    for (let i = 0; i < str.length; i++) {
+      hash = ((hash << 5) - hash) + str.charCodeAt(i); // essentially convert the number to binary, shift the bits over 5, and add the new number
+      hash |= 0; // converts the number to a 32bit integer
```
Not familiar with `|=`

micah-wehrle replied:
I had to look it up, because that whole dealio there I found online. It is a "bitwise or operator". So if you have two binary numbers, say 26 and 12, in binary they would be
`26 -> 11010`
`12 -> 01100`

So if we do `const result = 26 | 12` it would compare each bit of the above numbers and this would be the `result`:
`11010`
`01100`
`-----`
`11110`
So start with the ones column. Are 0 OR 0 "true"? No, so the result is 0. Then the tens column. Are 1 OR 0 "true"? Yes, so the result is 1. and so on. The answer 30.

The above thing, `num |= 0` essentially allows you convert whatever wild number is stored in `hash` back into an integer. Don't fully understand how it works but I'm guessing it's more efficient than whatever pleb way we would do it.

looksliketyler replied:
I'll have to check our code but I don't think we use it. 

Typically the use of helpful operators like this has to be agreed upon and put out as new standard. 

Add a comment above it showing an alternate approach.

micah-wehrle replied:
I just removed it. I don't know why we would need a 32 bit integer, and also I'm pretty sure the strings we're using won't be anywhere near 32 bits when converted to a number. I understand how the algorithm works, and this isn't needed.

looksliketyler commented on src/modules/jobs/models/tech-load.model.ts:
```diff
@@ -0,0 +1,399 @@
+
+  /**
+   * @description Will return a random element of the given array based off of the global seeded random number generator
+   * @param {string[]} options An array of possible options to pseudo-randomly choose from 
```
Doesn't match below

micah-wehrle replied:
Changed to T[]. I assume that's how you'd document the use of a generic?

looksliketyler commented on src/modules/jobs/models/tech-load.model.ts:
```diff
@@ -0,0 +1,399 @@
+    const firstName = this.generateFirstName();
+    const lastName = this.generateLastName();
+    const streetAddress = this.generateStreetAddress();
+    const jobType = this.randOf(['Install', 'Repair', 'Helper', 'POTS', 'BSW']);
```
randOf<T>()

You can pass type used here.  So 

`this.randOf<string[]>()`

micah-wehrle replied:
Wouldn't it be `this.randOf<string>()`? As the generic type is the same as the output, and the input is an array of the T. Also I didn't because it seems to infer it, but I will if you think it's good practice. Idk if it infers from the input or the variable into which the output is stored, but whatever the case if you hover over it: 
<img width="848" alt="Screen Shot 2023-05-20 at 10 59 45 PM" src="https://github.com/micah-wehrle/ABCNest/assets/11700686/229a692e-bea4-4899-a7ca-c4f139dcdffb">

looksliketyler replied:
Inference is best ts practice. Atlas standards using implicit typing. And yah. You're right. Drop the brackets.

micah-wehrle replied:
I changed to definite typing

looksliketyler commented on src/modules/jobs/models/tech-load.model.ts:
```diff
@@ -0,0 +1,399 @@
+  streetAddress: string,
+  state: string,
+  city: string,
+  zip?: number,
```
Put optional at end of interface.

micah-wehrle replied:
I moved lat/long to the top so I could keep city/state/zip together but have zip at the bottom.

looksliketyler commented on src/modules/jobs/models/tech-load.model.ts:
```diff
@@ -0,0 +1,399 @@
+interface Services {
+  equipment: string[];
+  internetSpeed?: number;
+  potsNumber?: number;
```
For reference, the API responses we get, numbers like these are typically strings.

StreamSG replied:
What's the benefit to that over having them as numbers?

StreamSG replied:
Still wanting to know @looksliketyler

looksliketyler commented on src/modules/jobs/models/tech-load.model.ts:
```diff
@@ -0,0 +1,399 @@
+   */
+  private generateCityLocation(streetAddress: string): Location {
+    const locations = `AL,Montgomery,32.377716,-86.300568;AK,Juneau,58.301598,-134.420212;AZ,Phoenix,33.448143,-112.096962;AR,Little Rock,34.746613,-92.288986;CA,Sacramento,38.576668,-121.493629;CO,Denver,39.739227,-104.984856;CT,Hartford,41.764046,-72.682198;DE,Dover,39.157307,-75.519722;HI,Honolulu,21.307442,-157.857376;FL,Tallahassee,30.438118,-84.281296;GA,Atlanta,33.749027,-84.388229;ID,Bois ...
+      .split(';').map(function(el: string): Location { // kinda gross, but this basically converts the above string into an array of City objects
```
Any reason using `function() {}` rather than arrow function?

micah-wehrle replied:
I think I could't figure out where to put the return type lol

StreamSG replied:
I've done this before hahaha

micah-wehrle replied:
I figured it out.

looksliketyler commented on src/modules/jobs/controllers/jobs.controller.ts:
```diff
@@ -1,14 +1,26 @@
+    if (!uuid || uuid.length === 0) {
+      return {
+        'flowStatus': 'FAILURE',
+        'flowStatusMessage': 'invalid UUID',
```
Return message is misleading. Says invalid uuid though the if statement is accounting for a lack of uuid...

I think you should add proper uuid formatting as the invalid option.

micah-wehrle replied:
It is done. Made a little better as well.

micah-wehrle committed:
added explicit typing
simplified job object creation
changed to arrow function in array map
moved optional variable to bottom of interface

micah-wehrle committed:
Addressed PR comments
made better flowStatusMessages
Better uuid validation
Adjusted jobs response formatting and class publicity

looksliketyler commented on src/modules/jobs/controllers/jobs.controller.ts:
```diff
@@ -1,14 +1,34 @@
+    if (!uuid || !uuid.match(/^[a-z]{2}\d{3}[a-z\d]$/i)) { // regex checks for valid uuid
+      return {
+        'flowStatus': 'FAILURE',
+        'flowStatusMessage': `${!uuid ? 'Please pass' : `${uuid} is not`} a valid uuid. Must be 6 characters: 2 letters, 3 digits, and then a letter or digit.`,
```
I'd keep this message short and simple.

StreamSG replied:
Might I suggest "Internal webphone Error"?

micah-wehrle replied:
Changed to just "invalid uuid"

looksliketyler commented on src/modules/jobs/controllers/jobs.controller.ts:
```diff
@@ -1,14 +1,34 @@
       'flowStatus': 'SUCCESS',
-      'flowStatusMessage': 'Successfully transmitted UUID.',
-      'data': uuid,
+      'flowStatusMessage': `Successfully generated job data from uuid ${uuid}`,
```
FlowStatusMessage is typically empty on a successful response for Atlas

micah-wehrle replied:
Emptied

looksliketyler commented on src/modules/jobs/models/tech-load.model.ts:
```diff
@@ -0,0 +1,398 @@
+
+  constructor(uuid: string, amount?: number) { // Can be built with a given amount, but if not will seed the number of jobs to build
+    this.uuid = uuid;
+    this.seed = Math.abs(this.convertStrToNum(uuid));
```
Add commenting explain the reason why Math.abs() being used

StreamSG replied:
@looksliketyler Wouldn't it be just making sure the number is positive?

micah-wehrle replied:
commented. it has to do with the fact that the seed increments++, so starting at a negative number can cause you to hit 0 which will always create the same number (because no matter what you multiply 0 by it'll always be 0)

looksliketyler commented on src/modules/jobs/models/tech-load.model.ts:
```diff
@@ -0,0 +1,398 @@
+    
+    this.jobs = [];
+    // generate a random number of jobs, weighted towards ~3
+    const generatedJobCount = Math.round( (this.nextRand()*7 + this.nextRand()*7) / 2 );  // This line needs to run regardless of if we were passed an amount, or else the seeded rng will get messed up!
```
Missing data type

micah-wehrle replied:
typed

looksliketyler commented on src/modules/jobs/models/tech-load.model.ts:
```diff
@@ -0,0 +1,398 @@
+      email: this.generateEmail(firstName, lastName),
+      phone: this.randDigits(10),
+      history: this.generateHistory(),
+      jobType: jobType,
```
```suggestion
      jobType
```

micah-wehrle replied:
simplified

looksliketyler commented on src/modules/jobs/models/tech-load.model.ts:
```diff
@@ -0,0 +1,398 @@
+      phone: this.randDigits(10),
+      history: this.generateHistory(),
+      jobType: jobType,
+      transportType: transportType,
```
```suggestion
      transportType
```

micah-wehrle replied:
simplified

looksliketyler commented on src/modules/jobs/controllers/jobs.controller.ts:
```diff
@@ -1,14 +1,34 @@
+    }
+
+    // Pass the amount whether it exists or not. TechLoad handles NaN (falsy) or an actual number
+    const jobData: TechLoad = new TechLoad(uuid, Number.parseInt(amount));
```
If amount is falsy, will Number.parseInt() throw an error?

micah-wehrle replied:
No, that's what the above comment is about, sorry I'll make it more clear. parseInt returns NaN if amount is undefined. The TechLoad constructor handles both a number and a falsy value

micah-wehrle replied:
Clarified better in the comment

micah-wehrle committed:
addressed PR comments
better documentation in a few areas
simplified code in creating JobData object literal

micah-wehrle committed:
added definite generic typing
